plugins/inputs/modbus/modbus.go

```diff
@@ -173,552 +167,253 @@ func (m *Modbus) Init() error {
-			}
+func (m *Modbus) gatherRequestsDiscrete(requests []request) error {
+	for _, request := range requests {
+		bytes, err := m.client.ReadDiscreteInputs(request.address, request.length)
+		if err != nil {
+			return err
 		}
 
-		if register.Type == cInputRegisters || register.Type == cHoldingRegisters {
-			for i := 0; i < len(register.Fields); i++ {
-				var valuesT []byte
-
-				for j := 0; j < len(register.Fields[i].Address); j++ {
-					tempArray := rawValues[register.Fields[i].Address[j]]
-					for x := 0; x < len(tempArray); x++ {
-						valuesT = append(valuesT, tempArray[x])
-					}
-				}
+		// Bit value handling
+		for i, field := range request.fields {
+			offset := field.address - request.address
+			idx := offset / 8
+			bit := offset % 8
 
-				register.Fields[i].value = convertDataType(register.Fields[i], valuesT)
-			}
+			request.fields[i].value = uint16((bytes[idx] >> bit) & 0x01)
 		}
 	}
-
 	return nil
 }
 
-func getBitValue(n byte, pos int) uint16 {
-	return uint16(n >> uint(pos) & 0x01)
-}
-
-func convertDataType(t fieldContainer, bytes []byte) interface{} {
-	switch t.DataType {
-	case "UINT16":
-		e16 := convertEndianness16(t.ByteOrder, bytes)
-		return scaleUint16(t.Scale, e16)
-	case "INT16":
-		e16 := convertEndianness16(t.ByteOrder, bytes)
-		f16 := int16(e16)
-		return scaleInt16(t.Scale, f16)
-	case "UINT32":
-		e32 := convertEndianness32(t.ByteOrder, bytes)
-		return scaleUint32(t.Scale, e32)
-	case "INT32":
-		e32 := convertEndianness32(t.ByteOrder, bytes)
-		f32 := int32(e32)
-		return scaleInt32(t.Scale, f32)
-	case "UINT64":
-		e64 := convertEndianness64(t.ByteOrder, bytes)
-		f64 := format64(t.DataType, e64).(uint64)
-		return scaleUint64(t.Scale, f64)
-	case "INT64":
-		e64 := convertEndianness64(t.ByteOrder, bytes)
-		f64 := format64(t.DataType, e64).(int64)
-		return scaleInt64(t.Scale, f64)
-	case "FLOAT32-IEEE":
-		e32 := convertEndianness32(t.ByteOrder, bytes)
-		f32 := math.Float32frombits(e32)
-		return scaleFloat32(t.Scale, f32)
-	case "FLOAT64-IEEE":
-		e64 := convertEndianness64(t.ByteOrder, bytes)
-		f64 := math.Float64frombits(e64)
-		return scaleFloat64(t.Scale, f64)
-	case "FIXED":
-		if len(bytes) == 2 {
-			e16 := convertEndianness16(t.ByteOrder, bytes)
-			f16 := int16(e16)
-			return scale16toFloat(t.Scale, f16)
-		} else if len(bytes) == 4 {
-			e32 := convertEndianness32(t.ByteOrder, bytes)
-			f32 := int32(e32)
-			return scale32toFloat(t.Scale, f32)
-		} else {
-			e64 := convertEndianness64(t.ByteOrder, bytes)
-			f64 := int64(e64)
-			return scale64toFloat(t.Scale, f64)
-		}
-	case "FLOAT32", "UFIXED":
-		if len(bytes) == 2 {
-			e16 := convertEndianness16(t.ByteOrder, bytes)
-			return scale16UtoFloat(t.Scale, e16)
-		} else if len(bytes) == 4 {
-			e32 := convertEndianness32(t.ByteOrder, bytes)
-			return scale32UtoFloat(t.Scale, e32)
-		} else {
-			e64 := convertEndianness64(t.ByteOrder, bytes)
-			return scale64UtoFloat(t.Scale, e64)
```

Comment:
looks like the behavior of this might have changed? Might not be covered by tests.

Reply:
Why do you think it has changed? The only difference is that this uses the number of bytes for determining the conversion and the new code uses the number of registers (=16bit). Furthermore, `TestHoldingRegisters` will check all those conversions.

One thing that changed though (for integers only) is that we only produce 64-bit output instead of 16, 32 or 64-bit variants for now in order to minimize the number of type conversion functions...

Reply:
did we drop support for `FLOAT64-IEEE` or any other types?

Reply:
No we didn't (take a look at the tests). The type conversions moved to separate files (`type_conversions*.go`).

In preparation of new configuration schemes I tried to unified the names because e.g. `FLOAT32`, `FLOAT32-IEEE`, `FIXED` and `UFIXED` caused a LOT of confusion. So the idea is to have a clear regime for future configuration schemes and still support the "old" configuration style. Therefore, the types are now normalized as `FLOAT32` & `UFIXED` to `UINTx` and `FIXED` to `INTx` (depending on the bit-size) in the configuration setup (see `normalizeInputDatatype` in `configuration_original.go`).